libnd4j/include/ops/declarable/platform/cudnn/cudnnUtils.h

```diff
@@ -60,9 +63,325 @@ namespace platforms {
+    operator cudnnFilterDescriptor_t() const { return desc; }
+
+    ~FilterDesc(){
+      if(desc_status==CUDNN_STATUS_SUCCESS)  cudnnDestroyFilterDescriptor(desc); ;
+    }
+
+    cudnnStatus_t desc_status;
+    cudnnFilterDescriptor_t desc;
+};
+
+struct DropoutDesc{
+
+    DropoutDesc(const DropoutDesc& s) = delete; 
+
+    DropoutDesc& operator=(const DropoutDesc& other) = delete;
+
+    DropoutDesc(){ 
+       create();
+    }
+
+    DropoutDesc(bool createDesc){ 
+       if(createDesc)  create();
+       else { desc= nullptr; }
+    }
+
+    void create(){
+      desc_status = cudnnCreateDropoutDescriptor(&desc); 
+    }
+
+    template<typename ...Args>
+    void set(cudnnStatus_t &last_err, Args&&... args){ 
+        if(desc_status==CUDNN_STATUS_SUCCESS && last_err==CUDNN_STATUS_SUCCESS){
+            last_err = cudnnSetDropoutDescriptor(desc, std::forward<Args>(args)...);
+            if(last_err){
+                nd4j_printf("Cudnn error code %s\n",cudnnGetErrorString(last_err));
+            }
+        }
+    }
+
+    cudnnStatus_t descStatus(){ return desc_status;}
+
+    operator cudnnDropoutDescriptor_t() const { return desc; }
+
+    ~DropoutDesc(){
+      if(desc_status==CUDNN_STATUS_SUCCESS)  cudnnDestroyDropoutDescriptor(desc); ;
+    }
+
+    cudnnStatus_t desc_status;
+    cudnnDropoutDescriptor_t desc;
+};
+
+#if CUDNN_VERSION > CUDNN_NEW_RNN_API_VER
+struct RnnDataDesc{
+
+    RnnDataDesc(const RnnDataDesc& s) = delete; 
+
+    RnnDataDesc& operator=(const RnnDataDesc& other) = delete;
+
+    RnnDataDesc(){ 
+       desc_status = cudnnCreateRNNDataDescriptor(&desc);
+    }
+
+    template<typename ...Args>
+    void set(cudnnStatus_t &last_err, Args&&... args){ 
+        if(desc_status==CUDNN_STATUS_SUCCESS && last_err==CUDNN_STATUS_SUCCESS){
+            last_err = cudnnSetRNNDataDescriptor(desc, std::forward<Args>(args)...);
+            if(last_err){
+                nd4j_printf("Cudnn error code %s\n",cudnnGetErrorString(last_err));
+            }
+        }
+    }
+
+    cudnnStatus_t descStatus(){ return desc_status;}
+
+    operator cudnnRNNDataDescriptor_t() const { return desc; }
+
+    ~RnnDataDesc(){
+      if(desc_status==CUDNN_STATUS_SUCCESS)  cudnnDestroyRNNDataDescriptor(desc); ;
+    }
+
+    cudnnStatus_t desc_status;
+    cudnnRNNDataDescriptor_t desc;
+};
+#endif
+
+FORCEINLINE cudnnStatus_t setRnnDescriptorOldApi(cudnnRNNDescriptor_t rnnDesc, 
+                            cudnnHandle_t handle,
+                            cudnnRNNInputMode_t inputMode,
+                            cudnnDirectionMode_t dirMode,
+                            cudnnRNNMode_t cellMode,
+                            cudnnRNNAlgo_t algo,
+                            cudnnDataType_t mathPrec,
+                            int32_t hiddenSize,
+                            int32_t numLayers,
+                            cudnnDropoutDescriptor_t dropoutDesc,  bool use_tensor_op=false){
+
+    cudnnStatus_t err = CUDNN_STATUS_SUCCESS;
```

Comment:
Mind adding some declarations here so we know what versions of methods are being invoked? nd4j_debug should be fine here

Reply:
I already wrote nd4j_debug that informs it will use v6 but it's in the caller function.

I tried to keep Scoped descriptors as simple as I can. Plus as history shows Cudnn side when upgrading they are not hesitant to change method signatures, introduce new functions, and so on. As they had newly introduced functions in v7 which were deprecated in v8. so I had to ignore some of them to make the code cleaner.
when CUDNN < 8.0.1 it calls v6 (CUDNN version 6) API ( but still allowing some functionality (for example clipping) by comparing against some versions).
for > 8 I'm using new v8 API calls. 
What's good with v8 they simplified/merged API calls, now most argument-based arrays are allocated on the device ( probably for the Cuda graph thing).

Reply:
The point was more to put this comment you just did in the comments so it's in the code and people don't have to search github to find context  :)